### ISSUE TITLE
Offer zip builds on Windows

### ISSUE THREAD

**Issue by ghost** [JuliaLang/julia#32158]
If you go here:

https://julialang.org/downloads

you see EXE installers for Windows. This contrasts with other programming
languages that offer both an executable installer and ZIP files that can be
extracted and run. For example Go:

https://golang.org/dl

and Nim:

https://nim-lang.org/install_windows.html

Python:

https://python.org/downloads/release/python-373

and C#:

https://dotnet.microsoft.com/download/dotnet-core/2.2

PHP:

https://windows.php.net/download

Perl:

http://strawberryperl.com/releases.html

**Comment by jebej** [comment]
It's not explicit, but you can open the exe file with e.g. 7zip.

**Comment by cup** [comment]
@jebej thank you. I am doing that currently, but its not ideal as Windows does not include 7zip

**Comment by ViralBShah** [comment]
Are these zip files signed in some way?

**Comment by cup** [comment]
@ViralBShah it would be the same process as this:

https://julialang-s3.julialang.org/bin/linux/x64/1.1/julia-1.1.1-linux-x86_64.tar.gz

except as a ZIP file

**Comment by stevengj** [comment]
What problem would this solve?

**Comment by cup** [comment]
@stevengj it would allow people to use the Julia environment without running an installer. If someone is a new user they may want to try Julia out, and installers can be invasive as they commonly:

- copy files to install dir
- copy files to profile dir
- set registry entries
- set environment variables

and historically uninstallers have a bad record of reverting these changes. should not put this burden upon new users or even having them resort to virtual machine to avoid items above. having a ZIP file allows new users to run the Julia environment, then if they like it perhaps they can run the installer going forward. Are you familiar with Windows? As everything I am describing hasnt changed in at least 20 years.

**Comment by hochB** [comment]
Would this allow people with no admin rights to run Julia on windows?

**Comment by ViralBShah** [comment]
Julia does not need admin rights even today, I believe.

**Comment by StefanKarpinski** [comment]
Does the Windows installer need admin rights to install? If it's just a matter of doing the same thing as the Linux tarball but for the Windows build and creating a zip file instead of a tarball, that would be easy enough and would be a great contribution, @cup.

**Comment by cup** [comment]
@StefanKarpinski and all, no, a ZIP file would not require any special privileges. Any user could download the ZIP file, extract, and run the Julia environment.

This is similar to Linux where a user could download a TAR.GZ, extract and run the executables. If you have a Windows environment, you can test this by using any of the links I put in the original post. Thank you!

**Comment by StefanKarpinski** [comment]
I was asking if the current Windows installer requires any special privileges, I don't believe it does, but I don't use Windows, so I'm not sure.

> This is similar to Linux where a user could download a TAR.GZ, extract and run the executables. If you have a Windows environment, you can test this by using any of the links I put in the original post. Thank you!

To be clear, I'm not offering to do this, nor am I volunteering anyone else to do it. If you want this feature, the best way to make it happen is to implement it and make a PR. Otherwise this issue is very likely to sit here open until someone else both wants this feature and is willing to implement it.

**Comment by cup** [comment]
@StefanKarpinski thanks for clearing it up. I am in a similar situation as you so it appears this will be sitting for some time.

I am ok with this, hopefully myself or someone will find some free time at one point to implement.

I would only add as a final note. I would be in the Julia group interest to implement this, as some might be avoiding the language because of this reason. Thanks again.

**Comment by visr** [comment]
The Windows installer does not need Administrator rights to install, only if you want to install it system wide.
https://github.com/JuliaLang/julia/blob/8f9d35627f5d01fb43fb325259bbea617741f191/contrib/windows/build-installer.nsi#L65

**Comment by StefanKarpinski** [comment]
@cup, given that you don't use Windows, can I ask what the motivation for this issue is?

**Comment by cup** [comment]
@StefanKarpinski not sure where you are getting that information, I use Primary Windows but also Linux via virtual machines

**Comment by StefanKarpinski** [comment]
> not sure where you are getting that information, I use primarily Windows but also Linux via virtual machines

I assumed that when you wrote:

> I am in a similar situation as you so it appears this will be sitting for some time.

that you meant you also don't run Windows and are therefore unable to implement and test this feature. Otherwise I don't understand what "similar situation" you are referring to.

**Comment by cup** [comment]
@StefanKarpinski 

> I'm not offering to do this, nor am I volunteering anyone else to do it.

that applies to me too. not that i dont want to implement it but i currently dont have time to. if that mean this issue doesnt get resolve right away so be it. im not asking or expecting anyone to fix it. obviously if it gets fixed i will be happy about that or maybe i will get around to it before anyone else.

**Comment by cup** [comment]
@StefanKarpinski 

> I'm not offering to do this, nor am I volunteering anyone else to do it.

that applies to me too. not that i dont want to implement it but i currently dont have time to. if that mean this issue doesnt get resolve right away so be it. im not asking or expecting anyone to fix it. obviously if it gets fixed i will be happy about that or maybe i will get around to it before anyone else.

i will only add that your tone is strange. are you saying that issues are only welcome if a user personally commits to implementing the fix? i think thats what sets aside a PR from an issue. again i am not against fixing this myself but until that time comes i think its constructive to have posted this anyway incase as you said someone else can fix it and has the time and/or desire to do so.

**Comment by StefanKarpinski** [comment]
Issues are welcome. My tone was in direct response to your apparent expectation that someone involved in the project should or would implement this. It seems that we’re in agreement that this can be tagged “help wanted” and left open.

**Comment by stevengj** [comment]
@cup, I've used the Windows installer; IIRC it just unpacks the files into a `julia-x.y` directory without modifying anything else if you don't check the box to install a Start-menu entry.  I don't think it modifies the registry or sets environment variables?

**Comment by cup** [comment]
@stevengj yes thats exactly right.

I dont have a criticism of the Julia installer, I have not run tests to be able
to determine if its well behaved or not. As you said, new users would have no
way to know this without doing their own testing, unless its behavior was
detailed on the download page itself or somewhere prominent.

So its the *not knowing* that is the issue. With Windows, every installer could
be perceived as a "black box". You mentioned another one - start menu entries,
so that makes at least 5 places that could be impacted:

- program files directory
- home directory
- registry entries
- environment variables
- start menu entries

It is comforting to know that Julia installer is well behaves, but I think it
still leaves the large issue unresolved. Its a tough problem because my instinct
says to keep it simple and only distribute one file.

If you distribute the ZIP only then its not surprises but you lose the
convenience of having start menu entries and PATH set automatically. If you
distribute the EXE only then you gain that convenience but maybe you lose users
who are wary of EXE files. I have also seen some distribute ZIP only with stub
installer in the top level directory which is a good solution but might be
confusing to novice users. However anyone looking to install Julia is prob not
novice. Finally you can just distribute both ZIP and EXE, which is what many
programming language vendors currently do, as shown in original post.

**Comment by davidanthoff** [comment]
> I've used the Windows installer many times; IIRC it just unpacks the files into a julia-x.y directory without modifying anything else if you don't check the box to install a Start-menu entry.

I think it does add itself to the uninstall list in control panel, right?

**Comment by ViralBShah** [comment]
I think a major motivation for this is that a lot of people don't have admin at work or school. Since we keep all our stuff in one folder, this should be relatively easy to do. 

The hard part is updating all the release processes.

**Comment by musm** [comment]
@cup just use the installer and run from the command line to extract to your desired folder. This is 100% portable.   On Julia 1.4 it's simply `julia-installer.exe /DIR=C:\temp `

I think this issue can be closed.

**Comment by cup** [comment]
@musm that would install Julia...

**Comment by musm** [comment]
No it doesn't it just extracts it.

**Comment by cup** [comment]
If that's the case, it should be documented somewhere. I think it's reasonable to expect that running something-installer.exe, is going to install that thing.

**Comment by ViralBShah** [comment]
Good point. It is reasonable that a .exe is an installer. We should document this.

**Comment by cup** [comment]
Also I havent mentioned it until now, but as I am getting some pushback it seems
important.

It looks like someone had made an error with the Windows packaging. Even if you
take `julia-1.2.0-win64.exe` or similar and extract it with 7-Zip or similar,
you dont get the actual distribution files, you get *another* installer,
`julia-installer.exe`.

Its only after you extract this *second* installer that you find the program
files. So even if this issue is closed as WONTFIX or similar, that separate
issue should be addressed.

**Comment by ViralBShah** [comment]
@musm mentioned that this only applies to Julia 1.4 (nightlies). However the point is that once you install into a specific directory, the resulting install is portable. I'll make a note in platform instructions on the downloads page.

**Comment by cup** [comment]
@ViralBShah @musm nope. Going here:

https://julialang.org/downloads/nightlies

and running

~~~
julia-latest-win64.exe /DIR=C:\temp 
~~~

installs it. You can tell because Julia now shows up under "Currently installed
programs" in the Windows Control Panel, and who knows where else. This is
exactly what I am trying to avoid.

**Comment by ViralBShah** [comment]
As we have repeatedly said, there is no other "where else". Everything is only installed in one folder. You can relocate it or copy it wherever you want. You can zip up that folder should you want to. I have noted so on the platform specific instructions page: https://julialang.org/downloads/platform.html

As you point out, we can provide a zip file for this as well. But for that, someone needs to make a contribution to the build systems and various scripts. We will leave this issue open for that.

**Comment by cup** [comment]
@ViralBShah can you clarify this?

> Uninstallation is performed by deleting the extracted directory and the
> packages directory in `%HOME%/.julia`. If you would also like to remove your
> preferences files, remove `%HOME%/.juliarc.jl` and `%HOME%/.julia_history`.

It doesnt account for the entry in the Control Panel.

**Comment by ViralBShah** [comment]
It's updated to refer to using the windows installer now. Please open separate issues or pull requests - and let this issue only be about the issue of zip file builds.

**Comment by musm** [comment]
@cup Try using http://innounp.sourceforge.net/  with the latest binaries. It looks like this tool will just extract the files without installing them as you want.

**Comment by cup** [comment]
@musm confirmed, thanks. Those are good workarounds.

**Comment by musm** [comment]
https://github.com/JuliaLang/julia/pull/35059#issuecomment-597905472

**Comment by musm** [comment]
We now have tarballs of portable Windows builds available on https://julialang.org/downloads/ 

Not exactly zip builds, but these should be easy enough to be extracted on Windows in the meantime.  We  should still probably  look at offering zip builds if it's much easier to use and if there's a strong preference for them.
